add preview proposal of new api for core

kodebach opened the pull request:
NOT INTENDED FOR MERGING

This PR is intended to discuss the ideas from #4761.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+
+- Collection: `KeySet`
+- Element: `Key`
+- Name of element: —
```
```suggestion
- Name of element: `char *`
```

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+- Collection: `ElektraSet` or `ElektraMap`
+- Element: `ElektraEntry`
+- Name of element: `ElektraKey` or `ElektraName`
+- Value of element: `ElektraValue`
```
I don't think this is a good idea. This would need a convincing usecase.

kodebach replied:
I don't quite understand the problem here... Basically what this say is:
1. The term for "the value of an Element" is: _value_
2. If we need to represent it as a type it is: `ElektraValue`

We might not use `ElektraValue` in C (let's discuss that in the structs section below), but in some bindings it could make sense. For example in Haskell:

```haskell
module Elektra (
  Value
) where

import Data.ByteString

newtype Value = Data.ByteString
```

kodebach replied:
Again. This part here is not about introducing `ElektraValue` as a type or identifier **anywhere**. This text is basically just a preview for the glossary. It just says we call this thing "value" (no other special name). The `ElektraValue` part basically just says **if** we found a reason to introduce a type, `ElektraValue` would be time. Essentially we reserve the name for this purpose.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+
+Another (maybe less confusing) option would be:
+
+- Collection: `ElektraSet` or `ElektraMap`
```
:+1: ElektraSet is nearer to current name, I don't see a reason to change to Map.

kodebach replied:
Well "Map, Entry, Key, Value" would be a pretty standard combination of terms.

Although changing the meaning of "Key" like that will be confusing for a while. It also means we actually need to update **all the docs** in more or less one go, instead of adding a notice "we changed terms from X to Y" somewhere and updating over a longer period of time.

If we want to use "Set", I actually prefer the combination: "Set, Entry, Name, Value".

markus2330 replied:
> instead of adding a notice "we changed terms from X to Y" 

This doesn't work anyway (even with FLOSS it would never be finished).

> If we want to use "Set", I actually prefer the combination: "Set, Entry, Name, Value".

Probably better to avoid `Key` anyway, see https://github.com/ElektraInitiative/libelektra/pull/4805#discussion_r1066061990

kodebach replied:
> This doesn't work anyway (even with FLOSS it would never be finished).

It is not intended as a permanent solution. But exactly because of FLOSS/CM having such a notice might be useful. Otherwise people could get confused, while the change happens, or we are limited to making the change when there is no FLOSS/CM happening.

Also we could always accidentally miss an occurrence in the update. The notice could also be put into the release notes. If we reuse "Key" we end up with something like:

```
Key was renamed to Entry and Keyname was renamed to Key.
```

Which is ambiguous. You can't tell whether "Key" is the old meaning and we forgot to update, or it is the new meaning.

markus2330 replied:
Back to topic: can you maybe create a second PR with this new naming schema? Maybe the look&feel gives a clearer picture.

I think "Key" -> "Entry" is a bit like our ref counting: a historical "problem", where the "problem" is that we do it differently then the rest of the world.

So these are border-line decisions (not very convincing problem but also not totally unconvincing): we should only tackle them if you are sure you can actually fully migrate&fix everything.

kodebach replied:
> we should only tackle them if you are sure you can actually fully migrate&fix everything

Renaming in C code is not an issue. Renaming in most other languages should be doable too. The issue is with documentation. I fear that has to be done manually. 

> I think "Key" -> "Entry" is a bit like our ref counting:

The difference is
1. ref counting is a technical issue with pros and cons on both sides
2. ref counting is a somewhat advanced topic that most people shouldn't need to worry about.

Using "Key" for something that is clearly not just a key, meanwhile, is just plain confusing to people who don't already know.

> Can you maybe create a second PR with this new naming schema?

I'll create a self-contained commit in this PR. If we like it we keep, if not I'll just revert the commit. Creating a new PR, unnecessarily splits the conversation. (It's unlikely the discussion in the new PR would really just focus on the names and nothing else).

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+Another (maybe less confusing) option would be:
+
+- Collection: `ElektraSet` or `ElektraMap`
+- Element: `ElektraEntry`
```
:+1: This is a huge change but probably a good idea as the "Key" for a key/value/meta was very confusing.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+
+- Collection: `ElektraSet` or `ElektraMap`
+- Element: `ElektraEntry`
+- Name of element: `ElektraKey` or `ElektraName`
```
Probably `ElektraKey` if we keep KDB.

kodebach replied:
We could go with EDB for ElektraDatabase. Would be another huge change, but no bigger than "Key -> Entry" really.

markus2330 replied:
I don't have troubles with EDB¹ and I don't think that the EDB change is as difficult, as *every* occurrence from KDB would be EDB afterwards. The trouble with Key is that there are many occurrences of Key not having the meaning of "ElektraEntry". But you probably know that very well anyway.

¹ historically there was the idea that "Elektra" is only a marketing name but not used in technical context. But I think we are already way beyond that point.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+
+### Note on public structs
+
+The public structs `ElektraKeyname` and `ElektraKeyvalue` defined below are not problematic or limiting in terms of forward compatibility.
```
I disagree. E.g. order or if namespace is separate should be implementation detail (only shared with mmap).

Anyway, we should have a decision about why we have opaque pointers, this is a common question and not so well known.

kodebach replied:
1. Whether namespace is separate or not is IMO actually decision we need to make. IMO (and I've stated that before), namespace and name should be entirely separate things. Now that I have thought more about the API, I actually don't see **any benefit** in storing the namespace as the first byte of the name field. IMO it just complicates things.

2. Yes, field order could be reason to make the struct opaque. But in this case there will always be the same 3 fields, a pointer, a size and a byte (for namespace). I don't see why we would ever want to change the order. In any case, I don't think field order outweighs all the benefits that having these public structs brings.

> we should have a decision about why we have opaque pointers, this is a common question and not so well known.

Then **you** need to write one. You are the only one who has been with Elektra long enough to know the actual reason.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+// 0 always indicates success, other values depend on function
+typedef int ElektraErrorCode;
+
+// Note: breaking change compared to current values
```
Not really relevant for pre-1.0

```suggestion
```

kodebach replied:
Yeah, I know I just added at for completeness's sake.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+typedef struct ElektraKey ElektraKey;
+typedef struct ElektraKeyset ElektraKeyset;
+
+// 0 always indicates success, other values depend on function
```
API usage would imho much easier if more than that is known.

kodebach replied:
IMO we should have rules/guidelines somewhere in `doc/contrib` so that our error codes are mostly uniform. So for example, "Illegal null pointer is -1". I just wouldn't document that in the user/developer facing headers, because then we'd make it an actual guarantee. Better to leave the option for some weird edge case function in a library somewhere that needs to use non-standard error codes.

kodebach replied:
PS. I also thought I about having some predefined values, that could be but don't have to be used by functions. That would just end up like `errno`, where every `E*` constant has a definition, but you (often) still need to check the individual function to see what the error actually means.

Again, we could have those constants in private headers to simplify writing code, but I wouldn't make that public.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+    ELEKTRA_KEY_CP_ALL    = ELEKTRA_KEY_CP_NAME | ELEKTRA_KEY_CP_VALUE | ELEKTRA_KEY_CP_META,
+} ElektraKeyCopyFlag;
+
+void elektraKeynamePushPart (ElektraKeyname * name, const char * part);
```
Imho add/append makes more sense.

kodebach replied:
Yes, "append" makes more sense. But there is no good single word counter part for removing. We could use "pop", but IMO "push"/"pop" as a pair makes more sense. Elektra is the only place where I've seen "append"/"pop" used a pair.

I didn't go with "add"/"remove", because that looses the information that we're only manipulating the last part. It would have to be the longer:

```c
void elektraKeynameAddLastPart (ElektraKeyname * name, const char * part);
// or maybe: void elektraKeynameAppendPart (ElektraKeyname * name, const char * part);
void elektraKeynameRemoveLastPart (ElektraKeyname * name);
```

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+const ElektraKeyname * elektraKeyGetName (const ElektraKey * key);
+// Note: creates copy of name
+ElektraErrorCode elektraKeySetName (ElektraKey * key, const ElektraKeyname * name);
+// increments nameLock
+void elektraKeyLockName (ElektraKey * key);
+// decrements nameLock
+void elektraKeyUnlockName (ElektraKey * key);
```
I am not a fan of having two ref counters. Imho it is over-engineering. Needs use case and decision.

kodebach replied:
**PLEASE** stop thinking of this as "two ref-counters". Describing it as that was the biggest mistake I/we ever made.

The name is locked with a re-entrant lock instead of a single bit flag.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+void elektraKeyLockName (ElektraKey * key);
+// decrements nameLock
+void elektraKeyUnlockName (ElektraKey * key);
+bool elektraKeyIsNameLocked (const ElektraKey * key);
```
We are still missing any indication if we should take care of having a small amount of functions. Please make such a benchmark. (e.g. size change if getting rid of all *Is* functions)

kodebach replied:
I have no idea what kind of benchmark you want here....

Also, how would you replace `elektraKeyIsNameLocked`? You can emulate it with other functions of `libelektra-core`.

Is the question just, whether to keep it in `libelektra-core` or to put it into `libelektra-extra`? If so, I really don't think right now is the time to discuss this.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+ElektraErrorCode elektraKeySetValue (ElektraKey * key, const ElektraKeyvalue * value);
+
+/**
+ * @returns a `const ElektraKeyset *` that can safely be cast to `ElektraKeyset *`, iff a non-const `ElektraKey *` was passed as @p key
```
?

kodebach replied:
The constness works like for e.g. `strchr`:

```c
const char * cs;
char * s;

// OKAY
const char * x = strchr(cs);
char * x = strchr(s);
const char * x = strchr(s);

// ILLEGAL CAST, non-const pointer into const string
char * x = strchr(cs);
```

Basically what it means that, this is correct API use:

```c
ElektraKey * key;

ElektraKeyset * meta = (ElektraKeyset *) elektraKeyGetMeta (key);
```

But this is still wrong, because the metadata of a const `Key` should also be const:

```c
const ElektraKey * constKey;

ElektraKeyset * meta = (ElektraKeyset *) elektraKeyGetMeta (constKey);
```

> **Note** I didn't go with the "const arg, non-const return" variant from `strchr`, because the illegal case with `strchr` is not detected by the compiler (the illegal cast would actually be within `strchr`). With `elektraKeyGetMeta` the caller needs to make the cast explicitly otherwise there will be a compiler warning/error. Also I believe, the main reason why `strchr` uses `const` the way it does, is because the original API was written before `const` existed in C and changing it now would be a breaking change.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+ */
+ELEKTRA_CONST ElektraKeyset * elektraKeyGetMeta (const ElektraKey * key);
+
+ElektraErrorCode elektraKeyIsBelow (const ElektraKey * first, const ElektraKey * second);
```
This looks weird as if only an error code is returned.

```suggestion
ElektraReturnCode elektraKeyIsBelow (const ElektraKey * first, const ElektraKey * second);
```

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+}
+
+ElektraKey * elektraKeysetGet (ElektraKeyset * ks, size_t index);
+ElektraKeyset * elektraKeysetGetRange (ElektraKeyset * ks, size_t start, size_t end);
```
This would need some use case.

kodebach replied:
See #4797. It is the basis for the cut functions described below.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+ElektraKeyset * elektraKeysetGetRange (ElektraKeyset * ks, size_t start, size_t end);
+
+void elektraKeysetRemove (ElektraKeyset * ks, size_t index);
+void elektraKeysetRemoveRange (ElektraKeyset * ks, size_t start, size_t end);
```
I don't think this is minimal API.

kodebach replied:
Can you give a reason?

If you want we can put the two `*Range` functions into `libelektra-extra`.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+
+size_t elektraKeysetLookup (ElektraKeyset * ks, const ElektraKeyname * name);
+
+size_t elektraKeysetFindHierarchy (ElektraKeyset * ks, const ElektraKeyname * root, size_t * end);
```
Your favorite function :wink: 

Is the `Find` necessary?

```suggestion
size_t elektraKeysetHierarchy (ElektraKeyset * ks, const ElektraKeyname * root, size_t * end);
```

kodebach replied:
> Your favorite function :wink:

The function is simply much more useful than `ksCut` IMO. It also makes a lot of sense with the `*Range` functions.

> Is the `Find` necessary?

I don't like function names without a verb. If I read code that uses APIs that I don't know, I find it much easier to understand, if every function name has a verb that tells me what's happening.

In this case, "find"  IMO indicates:
1. The operation is non-destructive.
2. There is actual work happening. Unlike a "get", a "find" indicates to me that I shouldn't call the function more than I need to and should rather cache results. As a concrete example, calling `elektraKeyGetName` multiple times instead of creating a local variable won't have much real world difference. But calling `elektraKeysetFindHierarchy` (on the same unchanged keyset) over and over again, will waste CPU time.
3. It is similar to a "lookup", but not quite the same.

kodebach replied:
See https://github.com/ElektraInitiative/libelektra/pull/4797/commits/1f2e0b902011e4a37e683749520eb9f258583e76 in #4797

markus2330 commented on doc/contrib/new_api.md:
````diff
@@ -0,0 +1,510 @@
+
+```c
+// Create key (Note: better ways below via additional APIs)
+ElektraKey * k1 = elektraKeyNew (&(ElektraKeyname){
+    .ns = ELEKTRA_NS_SYSTEM,
+    .name = "foo\0bar\0baz",
+    .size = 14
+});
+
````
Looks terrible.

kodebach replied:
That's a very personal opinion. I actually find it much more pleasing to look at than the current var-arg stuff in `keyNew`. Not least, because this has a very clear way of formatting it properly.

markus2330 replied:
I thought also rough and unspecific feedback is welcomed. I don't understand the reply, where exactly should I look?

kodebach replied:
> I thought also rough and unspecific feedback is welcomed

Yes, "looks terrible" is not feedback at all. I can't make use of this in anyway. I don't know which part you don't like, or what you'd like better.

> I don't understand the reply, where exactly should I look?

I don't understand the question...

What I personally said was, I find this `elektraKeyNew` call more less terrible (i.e. "more pleasing to look at") than the current `keyNew (..., KEY_VALUE, ... KEY_META, ...)` stuff.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+// manipulate name
+elektraKeynamePushPart (&name2, "boo");
+ElektraKey * k3 = elektraKeyNew (&name2);
+
```
example too long, can you please split up to small examples that show specific features of the API?

kodebach replied:
That would just repeat the setup over and over again. I know the example is long, but that's why I added so many comments...

markus2330 replied:
You can assume the setup to be done, or (I would prefer) compilable C in separate files where you can have a function/macro for common setup.

In general I think it would be a good idea if the whole API and examples are compilable (maybe even implemented to some degree, e.g. for ref counting). This would lower the barrier to actually try out modifying your examples.

kodebach replied:
Everything should compile already. Just put the API definition in a header (or into your test `.c` file) and try writing code. The _linker_ will of course complain because the functions aren't defined anywhere, but compiling should work...

More than that is not on the table right now. I won't invest the time to create an implementation that might not be used at all.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+ */
+ElektraErrorCode elektraKeynameProcessEscaped (
+    ElektraKeyname * dest,
+    const char * escaped,
+    char ** canonical,
+    size_t * canonicalSize,
+    const char ** errorPtr
```
I hope we can do better. (Unless too many functions are too much size-overhead.)

kodebach replied:
How do you want to do better? What is your problem with this function? Do you have any concrete ideas?

Just saying "I hope we can do better" helps nobody...

markus2330 replied:
We first need to benchmark the too-many-functions "problem". First the problem must be clear, then we can talk about solutions.

markus2330 replied:
The benchmark would be to compare:

- a version with very few functions and many arguments each, e.g. `elektraKeySetBit(int operation, int whatDoLock, ...)`
- a version with a minimal set of functions which have only little arguments each (more or less what we have now in the case of locking the key, asking for the sync bit etc.)

At the moment we are a bit inconsistent about this more functions vs. more arguments.

markus2330 replied:
No general conclusion needed. It is simply about what is better for Elektra, with the constraint that the core stays small in size.

markus2330 commented on doc/contrib/new_api.md:
```diff
@@ -0,0 +1,510 @@
+ElektraKeyset * ks = ELEKTRA_KEYSET (
+    // elektraKeysetBuild consumes reference, doesn't retain itself
+    // we call elektraKeyRetain, so we can keep our own reference
+    elektraKeyRetain (key), // refs = 2
```
Doesn't look easy to use.

kodebach replied:
Manual reference counting is never easy to use. Manual memory management in general is always hard to use. That's why all modern languages have some form of automatic memory management....

At least with this retain/release system it's clear when you should call retain/release. You retain, when you want a new reference for yourself that you hold. You release when you're done with that reference.

With the current inc/dec/del system it's just chaos. You inc when things broke and you dec as soon as possible. You call del when ever, because calling it too much doesn't matter. This sort of chaos is just cause trouble. It's also much easier to create memory leaks. If retain/release a segfault is much more likely than a memory leak and segfaults lead to crashes and issue reports. Memory leaks are silent and easily go undetected.

Anyway, the reference counting is the part of this proposal I'm least fixated on. I'm open to using any solution that works. I just used retain/release, because **I** like it best and I wanted to show that it is a good fit for Elektra (which has been questioned in the past).

kodebach replied:
> AFAIK we have less problems since keyDel is a NOP on wrong use. 

No it is not. "NOP on **wrong** use" is probably impossible and might be equivalent to solving the halting problem.

What you mean is that: Calling `keyDel` on a key with non-zero reference counter is a NOP.
But IMO that is not actually "wrong use" according to the current API and semantics.

In any case, wrong use definitely also includes a case like this:

```c
Key * k = keyNew (...);
keyDel (k);
keyDel (k);
```

And here obviously neither `keyDel` is a NOP.

> free(NULL) is also a NOP (maybe for that reason).

Please stop using this example, it is not a valid comparison. The equivalent for `keyDel` is `keyDel (NULL)`, which works exactly the same as `free (NULL)` (i.e. a NOP). The same would be true for `elektraKeyRelease (NULL)` which would also be a NOP.

> please let clarify such things only in decisions.

I tried in #4676, which went nowhere. Since you didn't even look at the solutions.

And like I said, the ref counting is the one part of the API proposal where I'm most open to other solutions. I just responded to once again, correct your misunderstandings/incorrect statements about the current `keyDel`.

markus2330 replied:
> What you mean is that

Important is what you mean, so there should be a very clear description of the problem in a decision.

> Please stop using this example, it is not a valid comparison.

If we are not open to different points of views, we will obviously struggle. I try to understand your viewpoint but failed up to now.

> Since you didn't even look at the solutions.

Because the described problem there (that we need ref counting) is already solved by the current API (=status quo). Decisions must be a problem of status quo, which might exist but is not described. Otherwise we only endlessly refactor based on different tastes.

kodebach replied:
> If we are not open to different points of views, we will obviously struggle.

I **am** open to different points of view. I'm just not open to ignoring or even accepting factually incorrect statements.

I have no pointed out to you on multiple occasions that your comparison of `keyDel` and `free` is flawed. You can compare `free (NULL)` and `keyDel (NULL)`. But you cannot compare `free (NULL)` with `keyDel (k)` where `k` is a key that's already deleted.

> > What you mean is that
> 
> Important is what you mean,

_No_, this was about what (I believe) **you** meant with "keyDel is a NOP on wrong use". You didn't mean "wrong use" in general, you just meant "calling keyDel on a key with non-zero reference counter". At least I assume that is what you mean, since it is the only NOP case for `keyDel` other than the trivial `keyDel (NULL)`.


> Decisions must be a problem of status quo, which might exist but is not described. Otherwise we only endlessly refactor based on different tastes.

Just as a sidenote, no idea to response: Once again, I started decisions for things that are not technically a problem with the status quo, because we are undeniably in the middle of a big API redesign. I already committed to doing that. I don't see why we should not at least think about doing more than strictly necessary, if it might result in further improvements. There is a "rejected" state for decisions. I don't understand, why you refuse to even start a discussion, when there is a clear option of "okay we explored our options, let's keep what we already have".

markus2330 reviewed (changes requested):
Many great ideas and improvements. I don't see a problem in merging this, especially if you add links to use cases and decisions, it would be a **great** overview.

kodebach commented:
@markus2330 Please respond to the comments I left. There are some answers I need from you, before I can work on this further. However, since you seem to be generally in favour of the changes, I see this as a lower priority than all the decision PRs.

> I don't see a problem in merging this, [...] it would be a great overview.

Maybe right now. But I don't see a point in keeping this, once the headers are updated and new examples are written.

markus2330 commented on doc/contrib/new_api/core/public.h:
```diff
@@ -0,0 +1,122 @@
+	ELEKTRA_KEY_CP_ALL = ELEKTRA_KEY_CP_NAME | ELEKTRA_KEY_CP_VALUE | ELEKTRA_KEY_CP_META,
+} ElektraEntryCopyFlag;
+
+void ElektraNamePushPart (ElektraName * name, const char * part);
```
I don't think we should use stack terminology for building up a path (rather append/add/concat/join or similar). Do we even need a "pop"-equivalent?

kodebach replied:
I used stack terminology, because we only support appending exactly one part to the end or removing exactly one part from the end. That is exactly LIFO behaviour. 

Like I said already, "append" is a good alternative for "push", but I don't see a good alternative for "pop" (*). In that case I much prefer using push/pop, a common pair, rather than e.g. append/pop, which may confuse people because it seems like a mismatch.

(*) "remove"/"delete"/etc. all convey removing something, but not that it is explicitly the last element that gets removed.

> Do we even need a "pop"-equivalent?

Well, if you can only add to a name, there would be no way to replace the current `keySetBaseName` (without rebuilding the entire name).

markus2330 reviewed (changes requested):
ElektraSet/ElektraEntry looks good to me. Looking forward to other opinions.

@kodebach you can also invite others to have a look after you changed the main problems: public structs, ref counting (or created decisions for it).

kodebach commented on doc/contrib/new_api/README.md:
```diff
@@ -0,0 +1,110 @@
+- Element: `ElektraEntry`
+- Name of element: `ElektraName`
+- Value of element: `ElektraValue`
+
+### Note on public structs
+
+The public structs `ElektraName` and `ElektraValue` defined here are not problematic or limiting in terms of forward compatibility.
+This is because, they essentially define what a "keyname" and a "key value" are for Elektra.
+That is why they are used both as part of the public API **and** in the definitions of the `*Cow` structs, which define how stuff is actually stored.
+Any change to these structs would mean changing that definition and therefore would be major breaking change, even if the structs where not public.
```
Now that I have explained my reasons, can you maybe explain why this is not convincing for you?

Can you give an example, where a breaking change is needed with this API

```c
const ElektraName * elektraEntryGetName (const ElektraEntry * entry);
```

but could be avoided with an API without structs like below?

```c
ElektraNamespace elektraEntryGetNamespace (const ElektraEntry * entry);
const char * elektraEntryGetName (const ElektraEntry * entry);
size_t elektraEntryGetNameSize (const ElektraEntry * enty);
```

kodebach replied:
A half-way option could be introducing only a public struct

```c
typedef struct {
    char * string;
    size_t size;
} ElektraString;
```

I think we can all agree that this is something that will never change. Then we can write the API like this:

```c
ElektraNamespace elektraEntryGetNamespace (const ElektraEntry * entry);
const ElektraString * elektraEntryGetName (const ElektraEntry * entry);

ElektraReturnCode ElektraSetNamespace (ElektraEntry * key, ElektraNamespace ns);
ElektraReturnCode ElektraSetName (ElektraEntry * key, const ElektraString * name);

const ElektraString * elektraEntryGetValue (const ElektraEntry * entry);
ElektraReturnCode ElektraSetValue (ElektraEntry * key, const ElektraString * value);
```

That way we still avoid all the separate `char *` and `size_t` arguments that could get mixed up.

> **Note**: If we go this way, I'll have to rethink whether passing by ref (`ElektraString *`) is still better than by value (`ElektraString`).

kodebach commented on doc/contrib/new_api/README.md:
````diff
@@ -0,0 +1,110 @@
+An "unsafe" operation, does not immediately break anything, but it is still an incorrect use of the API and often creates new "unsafe" or even "illegal" operations for the future.
+
+```c
+ElektraEntry * k1 = ElektraEntryNew (&(ElektraName){
````
1. The change to uppercase function names was by mistake. It seems I didn't enable "case preserving" when replacing key -> entry.
2. It's not so bad for this function, but for other functions you may get multiple `const char*` and especially `size_t` arguments. With a wrapper struct it's much clearer which arguments belong to each other. Also once again, these structs **are not internal**.

> 2 just counts the `\0`

Interesting idea, but I think that would be (1) exceedingly confusing to newcomers (2) makes the internals much less efficient. We'd need to iterate the string to find it's length (skipping over the given number of `\0`) and only then could we `memcpy` it. It also means you need an extra function (or field in `ElektraName`), which tells you the number of parts (`\0`s) in an entry's name.

kodebach commented on doc/contrib/new_api/builders/public.h:
```diff
@@ -0,0 +1,69 @@
+	}
+
+/**
+ * CAUTION releases reference to @p meta
```
Well it's what makes most sense with the refcounting scheme used here. If we don't consume the reference the examples in `builders/examples.c` would leak. You would have to build all the metadata sets in separate variables use those in the build calls and then release all the metadata sets afterwards. This defeats the point of the builders, which is to have a single function that builds an entire set.

kodebach commented on doc/contrib/new_api/builders/examples.c:
```diff
@@ -0,0 +1,24 @@
+{
+	int a = 7;
+
+	ElektraEntry * key = ElektraEntryNew (&ELEKTRA_KEYNAME (ELEKTRA_NS_SYSTEM, "foo")); // refs = 1
```
Again, uppercase names were accidental. I don't know what you mean with "several APIs for the same"?

kodebach commented:
@markus2330 I fixed the function names (now proper camelCase). I also added examples for the two reference counting schemes:

1. `ref_inc_dec.c` shows the scheme of the current API with `inc`/`dec`/`del`
2. `ref_retain.c` shows a retain/release scheme, where "insert" moves an existing reference into the set
3. `ref_retain2.c` shows retain/release, but with two "insert" variants. One that creates a new reference and another that creates a new internal reference, but also destroys the external reference.

The original proposal used option 3, which IMO is the simplest option from a usage POV. However, I fully agree that changing the refcounting scheme is out of scope for use. That's why the proposal now uses a variation of option 1. The only change compared to the current API is that `inc`/`dec` now return their argument to allow inline use.

> **Note**: I need to rethink what we do about `const` in the API. I still like the idea behind `ELEKTRA_CONST`, but it's a bit annoying to use.

kodebach commented:
I will reopen this PR in a clean state with a few different options for the new API.
